Work around for bias quantization for conv and linear operators

raghuramank100 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24924 Serialization for nn.quantized.functional modules
* #24923 Support observer without any data calibration
* #24804 Update mapping dictionary to support functionalmodules and pooling operations
* #24803 Handle empty qconfig for functional Modules
* **#24789 Work around for bias quantization for conv and linear operators**

In eager mode, all modules need to work with input tensors that can change qparams dynamically. This issue https://github.com/pytorch/pytorch/issues/23874 will address this via FBGEMM modifications. This is a work around before that.

Differential Revision: [D16852280](https://our.internmc.facebook.com/intern/diff/D16852280/)

raghuramank10000 committed:
Work around for bias quantization for conv and linear operators

In eager mode, all modules need to work with input tensors that can change qparams dynamically. This issue https://github.com/pytorch/pytorch/issues/23874 will address this via FBGEMM modifications. This is a work around before that.

Differential Revision: [D16852280](https://our.internmc.facebook.com/intern/diff/D16852280/)

raghuramank10000 committed:
Update on "Work around for bias quantization for conv and linear operators"


In eager mode, all modules need to work with input tensors that can change qparams dynamically. This issue https://github.com/pytorch/pytorch/issues/23874 will address this via FBGEMM modifications. This is a work around before that.

Differential Revision: [D16852280](https://our.internmc.facebook.com/intern/diff/D16852280/)

raghuramank10000 committed:
Update on "Work around for bias quantization for conv and linear operators"


In eager mode, all modules need to work with input tensors that can change qparams dynamically. This issue https://github.com/pytorch/pytorch/issues/23874 will address this via FBGEMM modifications. This is a work around before that.

Differential Revision: [D16852280](https://our.internmc.facebook.com/intern/diff/D16852280/)

dskhudia commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -214,7 +225,8 @@ def from_float(cls, mod):
         assert weight_observer.dtype == torch.qint8, 'Weight observer must have a dtype of qint8'
         wt_scale, wt_zp = weight_observer.calculate_qparams()
-        bias_scale = float(wt_scale * act_scale)
+        bias_scale = float(act_scale / (2**16))
```
Any reason to quantize bias using act_scale / 2**16 ? Wouldn't quantizing it with just act_scale work?

raghuramank100 replied:
This was done to use as much precision for bias as possible. If we use act_scale, bias would be quantized to 8 bits, instead of 32 bits.

dskhudia replied:
act_scale (aka output_scale) quantizes output_act to 8-bits. We assume that bias is on the same range as output so act_scale will quantize bias to 8-bit as well? Is this the reasoning? Just asking for my understanding.

raghuramank100 replied:
Yes, thats my reasoning. bias should be comparable to the raw conv outputs and therefore the output scale is a proxy for the bias scale itself

jianyuh commented:
Do we have the unit test coverage for this change?

raghuramank10000 committed:
Update on "Work around for bias quantization for conv and linear operators"


In eager mode, all modules need to work with input tensors that can change qparams dynamically. This issue https://github.com/pytorch/pytorch/issues/23874 will address this via FBGEMM modifications. This is a work around before that.

Differential Revision: [D16852280](https://our.internmc.facebook.com/intern/diff/D16852280/)

dskhudia commented:
This would cause failures for per channel quantized weights.

raghuramank10000 committed:
Update on "Work around for bias quantization for conv and linear operators"


In eager mode, all modules need to work with input tensors that can change qparams dynamically. This issue https://github.com/pytorch/pytorch/issues/23874 will address this via FBGEMM modifications. This is a work around before that.

Differential Revision: [D16852280](https://our.internmc.facebook.com/intern/diff/D16852280/)

raghuramank100 commented:
> This would cause failures for per channel quantized weights.
The work around does not address this case, for now we support only per tensor quantization in the modules.

raghuramank10000 committed:
Update on "Work around for bias quantization for conv and linear operators"


In eager mode, all modules need to work with input tensors that can change qparams dynamically. This issue https://github.com/pytorch/pytorch/issues/23874 will address this via FBGEMM modifications. This is a work around before that.

Differential Revision: [D16852280](https://our.internmc.facebook.com/intern/diff/D16852280/)

dskhudia commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -120,6 +120,7 @@ def __init__(self, in_features, out_features, bias_=True):
             [out_features, in_features], scale=1, zero_point=0, dtype=torch.qint8)
 
         self.set_weight(qweight)
+        self.weight_scale = 1.0
```
May be we can add a check here to make sure that the qscheme of weights is kPerTensorAffine.

dskhudia commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -106,6 +111,7 @@ def extra_repr(self):
     def set_weight(self, w):
         self._packed_weight = torch.ops.quantized.fbgemm_conv_prepack(
             w.permute([0, 2, 3, 1]), self.stride, self.padding, self.dilation, self.groups)
+        self.weight_scale = w.q_scale()
```
Check here to make sure that the qscheme of weights is kPerTensorAffine.

dskhudia commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -83,10 +83,15 @@ def __init__(self, in_channels, out_channels, kernel_size, stride=1,
             [out_channels, in_channels // self.groups, self.kernel_size[0],
                 self.kernel_size[1]],
             scale=1, zero_point=0, dtype=torch.qint8)
+        self.weight_scale = 1.0
```
May be we can add a check here to make sure that the qscheme of weights is kPerTensorAffine.

raghuramank100 replied:
Currently, perTensorAffine is the only quantization type, so this is not needed. We can add checks for perTensorAffine as we start implementing perChannelQuantization support in Conv2d as a later task.

dskhudia replied:
I meant adding an assert to error out explicitly if these modules are called with per channel quantization. Otherwise it might exit with some unrelated error.

raghuramank10000 committed:
Update on "Work around for bias quantization for conv and linear operators"


In eager mode, all modules need to work with input tensors that can change qparams dynamically. This issue https://github.com/pytorch/pytorch/issues/23874 will address this via FBGEMM modifications. This is a work around before that.

Differential Revision: [D16852280](https://our.internmc.facebook.com/intern/diff/D16852280/)

raghuramank10000 committed:
Update on "Work around for bias quantization for conv and linear operators"


In eager mode, all modules need to work with input tensors that can change qparams dynamically. This issue https://github.com/pytorch/pytorch/issues/23874 will address this via FBGEMM modifications. This is a work around before that.

Differential Revision: [D16852280](https://our.internmc.facebook.com/intern/diff/D16852280/)

raghuramank10000 committed:
Update on "Work around for bias quantization for conv and linear operators"


In eager mode, all modules need to work with input tensors that can change qparams dynamically. This issue https://github.com/pytorch/pytorch/issues/23874 will address this via FBGEMM modifications. This is a work around before that.

Differential Revision: [D16852280](https://our.internmc.facebook.com/intern/diff/D16852280/)
